### PR TITLE
solution

### PR DESCRIPTION
**Pull request by lefteris-L**

### CODE REVIEW

File: src/sortStudents.ts [hunk 1]
```diff
@@ -1,16 +1,65 @@
...
+    if (sortBy === SortType.AverageGrade) {
+      const gradesA: Num[] = studentA.grades;
+      const gradesB: Num[] = studentB.grades;
+
+      const getSum = (x: Num, y: Num): Num => x + y;
+
+      const averageA: Num = gradesA.reduce(getSum) / gradesA.length;
+      const averageB: Num = gradesB.reduce(getSum) / gradesB.length;
+
+      return averageA - averageB;
```
Comment (ysrckr) [inline src/sortStudents.ts:41]: You can make a function to calculate average grade so you wouldn't duplicate same things for each of them.

File: src/sortStudents.ts [hunk 2]
```diff
@@ -1,16 +1,65 @@
...
-  // write your function
-}
+  const sortCallback = (a: Student, b: Student): Num => {
+    let studentA: Student = a;
+    let studentB: Student = b;
+
+    if (order === 'desc') {
+      studentA = b;
+      studentB = a;
+    }
```
Comment (ysrckr) [inline src/sortStudents.ts:30]: You can write with ternary so you wouldn't need to reassign.

File: src/sortStudents.ts [hunk 3]
```diff
@@ -1,16 +1,65 @@
...
+      case SortType.Surname:
+        return studentA.surname.localeCompare(studentB.surname);
+
+      case SortType.Age:
+        return studentA.age - studentB.age;
+
+      case SortType.Married:
+        return !studentA.married && studentB.married
+          ? -1
+          : 1;
```
Comment (ysrckr) [inline src/sortStudents.ts:57]: If you change boolean to number, you wouldn't need the ternary.

**Review by ysrckr — CHANGES_REQUESTED** [review]

File: src/sortStudents.ts (hunk 3 above)
Comment (lefteris-L) [reply src/sortStudents.ts:57]: Great suggestions, thanks. Used all of them

File: src/sortStudents.ts [hunk 4]
```diff
@@ -1,16 +1,64 @@
...
+      : b;
+
+    const studentB: Student = order === 'asc'
+      ? b
+      : a;
+
+    const isMarried = (student: Student): Num => (student.married
+      ? 1
+      : 0
+    );
```
Comment (ysrckr) [inline src/sortStudents.ts:35]: You don't need this to convert `Boolean` to `Number`. 
You can write Number(studentA) to convert it to `Number`

**Review by ysrckr — APPROVED** [review]
LGTM 🤟🏻
Probably Nazar or Roman will ask for a few changes but I think this is good. 😄

File: src/sortStudents.ts (hunk 4 above)
Comment (lefteris-L) [reply src/sortStudents.ts]: Thanks, didn't know that

File: src/sortStudents.ts [hunk 5]
```diff
@@ -1,16 +1,59 @@
...
+  Surname,
+  Age,
+  Married,
+  AverageGrade,
 }
 
-// create SortOrder type
-export type SortOrder;
+export type SortOrder = 'asc' | 'desc';
+export type SortFunc = (a: Student[], b: SortType, c: SortOrder) => Student[];
```
Comment (meforoma) [inline src/sortStudents.ts:18]: really nice 👏

File: src/sortStudents.ts [hunk 6]
```diff
@@ -1,16 +1,59 @@
...
-  // write your function
-}
+  const sortCallback = (a: Student, b: Student): Num => {
+    const studentA: Student = order === 'asc'
+      ? a
+      : b;
+
+    const studentB: Student = order === 'asc'
+      ? b
+      : a;
```
Comment (meforoma) [inline src/sortStudents.ts:30]: you could make this to shorten the code
```suggestion
    const [studentA, studentB] = order === 'asc'
      ? [a, b]
      : [b, a];
```

File: src/sortStudents.ts [hunk 7]
```diff
@@ -1,16 +1,59 @@
...
+    const getAverageGradeOf = (student: Student): Num => (
+      student.grades.reduce(getSum) / student.grades.length
+    );
+
+    switch (sortBy) {
+      case SortType.Name:
+        return studentA.name.localeCompare(studentB.name);
+
+      case SortType.Surname:
+        return studentA.surname.localeCompare(studentB.surname);
```
Comment (meforoma) [inline src/sortStudents.ts:42]: these cases are essentially the same - let's join them
```suggestion
      case SortType.Name:
      case SortType.Surname:
        return studentA.surname.localeCompare(studentB.surname);
```

File: src/sortStudents.ts [hunk 8]
```diff
@@ -1,16 +1,59 @@
...
+
+      case SortType.Age:
+        return studentA.age - studentB.age;
+
+      case SortType.Married:
+        return Number(studentA.married) - Number(studentB.married);
```
Comment (meforoma) [inline src/sortStudents.ts:48]: same here - join cases
```suggestion
      case SortType.Age:
      case SortType.Married:
        return Number(studentA.married) - Number(studentB.married);
```

**Review by meforoma — CHANGES_REQUESTED** [review]
really un-trivial approach and that's awesome 👍 👍 👍 
let's just make the code bit shorter - take a look at few suggestions.
I'd actually approve already, but want to make sure you apply those optimizations 🙂

**Comment by lefteris-L** [comment]
> really un-trivial approach and that's awesome 👍 👍 👍 let's just make the code bit shorter - take a look at few suggestions. I'd actually approve already, but want to make sure you apply those optimizations 🙂

Absolutely, will do

File: src/sortStudents.ts (hunk 6 above)
Comment (lefteris-L) [reply src/sortStudents.ts:30]: Awesome

**Pull request closed without merging**